### DISCUSSION TITLE
Duckdb using incorrect data type when reading from parquet file

### DISCUSSION THREAD

**Discussion by in42** [duckdb/duckdb#12559] [Q&A]
Hi, I have a parquet file and when reading from it using duckdb I get an incorrect pandas data type as opposed to reading it directly using pandas, as shown below:
```
>>> duckdb.sql('select "dummy_column" from dummy_file.parquet').df().dtypes
status_change.trading    object                     
dtype: object                                       
>>> pandas.read_parquet("dummy_file.parquet")["dummy_column"].dtypes
CategoricalDtype(categories=['UNKNOWN', 'ON', 'OFF'], ordered=False, categories_dtype=object)
```
How do I get a dataframe with the correct column type when reading from the parquet file using duckdb?

**Comment by soerenwolfers** [comment]
What data type does duckdb use to represent thr parquet categorical column? (Put a DESCRIBE before your select, and don't run .df() on it).

**Reply by in42** [reply]
```
In [3]: duckdb.sql('describe select dummy_column from dummy_file.parquet')
Out[3]: 
┌──────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│ column_name  │ column_type │  null   │   key   │ default │  extra  │
│   varchar    │   varchar   │ varchar │ varchar │ varchar │ varchar │
├──────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ dummy_column │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
└──────────────┴─────────────┴─────────┴─────────┴─────────┴─────────┘
```

**Reply by soerenwolfers** [reply]
Right. So you can create the enum type yourself via

```
CREATE TYPE birds AS ENUM (SELECT dummy_column FROM dummy_file.parquet);
```
and then do 
```sql
select dummy_column::birds, * from dummy_file.parquet
```

Duckdb does automatically convert enum columns to pandas' categorical columns.

**Reply by in42** [reply]
Ok, this seems to work. Thanks, this was very helpful.

**Reply by in42** [reply]
One issue though, this line picks up values only present in that column:

```
CREATE TYPE birds AS ENUM (SELECT dummy_column FROM dummy_file.parquet);
```

So even if the original enum had a `OFF` value, if it does not appear in that column it is not picked up. Is there any way to do that?

I can manually define the values in the enum, however that will be tedious, if there are too many enum values.

**Reply by in42** [reply]
Also manually defining the enum values may make it out of sync with the parquet file.

**Reply by soerenwolfers** [reply]
Not sure what you mean by out of sync. To add values manually, just extend the query in the enum definition with `UNION`

**Reply by in42** [reply]
So basically the duckdb enum type is being created with the values that are only present in the column. For example, if in the parquet file, a category has 3 values `OFF`, `ON` and `UNKNOWN`. However, if only one of them actually appears in that column e.g. `OFF` (all the rows have value `OFF` for that column), then the duckdb enum type created has only one value `OFF`. Is there any way to pick up the enum values for the duckdb enum from the category definition, rather than from the values appearing in the column?

**Reply by in42** [reply]
I guess this is related: #9617

**Reply by soerenwolfers** [reply]
As other people have said, there is no categorical data type in parquet. Therefore you're going to have to define the enum values yourself, using any query you want. You might find the inofficial definition of the categories in the parquet metadata, which you can query with duckdb as described at https://duckdb.org/docs/data/parquet/metadata.html

**Reply by soerenwolfers** [reply]
And yes, following the link in your link does show that the key value metadata contains pandas' inofficial definition of the category columns, which you can use to piece it all together. Unfortunately you're going to have to do it yourself for now

**Reply by soerenwolfers** [reply]
You should up vote #9617 to help its prioritization by the way

**Comment by eitsupi** [comment]
See #3309
I don't think there are Categorical types in Parquet.

**Reply by Tishj** [reply]
I imagine they use the "metadata" field to convey that this is categorical
